File: dp/c/monitor_object.c
// Monitor Object Pattern in C
// Synchronized bounded buffer using mutex + condition variables

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

#define CAPACITY 4

typedef struct {
    int items[CAPACITY];
    int count, in_idx, out_idx;
    pthread_mutex_t mutex;
    pthread_cond_t not_empty;
    pthread_cond_t not_full;
} Monitor;

void monitor_init(Monitor* m) {
    m->count = m->in_idx = m->out_idx = 0;
    pthread_mutex_init(&m->mutex, NULL);
    pthread_cond_init(&m->not_empty, NULL);
    pthread_cond_init(&m->not_full, NULL);
}

void monitor_put(Monitor* m, int item) {
    pthread_mutex_lock(&m->mutex);
    while (m->count == CAPACITY)
        pthread_cond_wait(&m->not_full, &m->mutex);

    m->items[m->in_idx] = item;
    m->in_idx = (m->in_idx + 1) % CAPACITY;
    m->count++;
    printf("[Monitor] Put %d (count=%d)\n", item, m->count);
    pthread_cond_signal(&m->not_empty);
    pthread_mutex_unlock(&m->mutex);
}

int monitor_take(Monitor* m) {
    pthread_mutex_lock(&m->mutex);
    while (m->count == 0)
        pthread_cond_wait(&m->not_empty, &m->mutex);

    int item = m->items[m->out_idx];
    m->out_idx = (m->out_idx + 1) % CAPACITY;
    m->count--;
    printf("[Monitor] Take %d (count=%d)\n", item, m->count);
    pthread_cond_signal(&m->not_full);
    pthread_mutex_unlock(&m->mutex);
    return item;
}

void monitor_destroy(Monitor* m) {
    pthread_mutex_destroy(&m->mutex);
    pthread_cond_destroy(&m->not_empty);
    pthread_cond_destroy(&m->not_full);
}

static Monitor mon;

void* producer(void* arg) {
    for (int i = 0; i < 8; i++) {
        monitor_put(&mon, i);
        usleep(20000);
    }
    return NULL;
}

void* consumer(void* arg) {
    for (int i = 0; i < 8; i++) {
        int val = monitor_take(&mon);
        printf("  Consumed: %d\n", val);
        usleep(40000);
    }
    return NULL;
}

int main(void) {
    monitor_init(&mon);
    pthread_t p, c;
    pthread_create(&p, NULL, producer, NULL);
    pthread_create(&c, NULL, consumer, NULL);
    pthread_join(p, NULL);
    pthread_join(c, NULL);
    monitor_destroy(&mon);
    return 0;
}

File: dp/rust/src/monitor_object.rs
// Monitor Object Pattern in Rust
// Synchronized object with mutual exclusion using Mutex + Condvar

use std::sync::{Arc, Condvar, Mutex};
use std::thread;

struct BoundedBuffer<T> {
    data: Mutex<BufferInner<T>>,
    not_empty: Condvar,
    not_full: Condvar,
}

struct BufferInner<T> {
    items: Vec<T>,
    capacity: usize,
}

impl<T: std::fmt::Debug> BoundedBuffer<T> {
    fn new(capacity: usize) -> Self {
        BoundedBuffer {
            data: Mutex::new(BufferInner {
                items: Vec::with_capacity(capacity),
                capacity,
            }),
            not_empty: Condvar::new(),
            not_full: Condvar::new(),
        }
    }

    fn put(&self, item: T) {
        let mut buf = self.data.lock().unwrap();
        while buf.items.len() >= buf.capacity {
            buf = self.not_full.wait(buf).unwrap();
        }
        println!("[Monitor] Put: {:?}", item);
        buf.items.push(item);
        self.not_empty.notify_one();
    }

    fn take(&self) -> T {
        let mut buf = self.data.lock().unwrap();
        while buf.items.is_empty() {
            buf = self.not_empty.wait(buf).unwrap();
        }
        let item = buf.items.remove(0);
        println!("[Monitor] Take: {:?}", item);
        self.not_full.notify_one();
        item
    }
}

fn main() {
    let buffer = Arc::new(BoundedBuffer::new(3));

    let producer = {
        let buf = Arc::clone(&buffer);
        thread::spawn(move || {
            for i in 0..6 {
                buf.put(i);
                thread::sleep(std::time::Duration::from_millis(30));
            }
        })
    };

    let consumer = {
        let buf = Arc::clone(&buffer);
        thread::spawn(move || {
            for _ in 0..6 {
                let val = buf.take();
                println!("  Consumed: {}", val);
                thread::sleep(std::time::Duration::from_millis(60));
            }
        })
    };

    producer.join().unwrap();
    consumer.join().unwrap();
}
